ffi/derive/src/convert.rs

```diff
@@ -518,136 +838,152 @@ fn gen_discriminants(
     )
 }
 
-fn variant_discriminants(enum_: &DataEnum) -> Vec<proc_macro2::Literal> {
-    enum_
-        .variants
+fn variant_discriminants(variants: &[SpannedValue<FfiTypeVariant>]) -> Vec<proc_macro2::Literal> {
+    variants
         .iter()
         .enumerate()
         .map(|(i, _)| proc_macro2::Literal::usize_unsuffixed(i))
         .collect()
 }
 
-fn variant_mapper<F0: FnOnce() -> TokenStream, F1: FnOnce(&syn::Field) -> TokenStream>(
-    variant: &syn::Variant,
+fn variant_mapper<F0: FnOnce() -> TokenStream, F1: FnOnce(&FfiTypeField) -> TokenStream>(
+    emitter: &mut Emitter,
+    variant: &SpannedValue<FfiTypeVariant>,
     unit_mapper: F0,
     field_mapper: F1,
 ) -> TokenStream {
-    match &variant.fields {
-        syn::Fields::Unnamed(syn::FieldsUnnamed { unnamed, .. }) if unnamed.len() == 1 => {
-            field_mapper(&unnamed[0])
-        }
-        syn::Fields::Unnamed(syn::FieldsUnnamed { unnamed, .. }) => {
-            abort!(unnamed, "Only 1-sized variants are supported")
+    match &variant.fields.style {
```

Comment:
This match is identical to the one which is above **in terms of validation and error messages**. Could you peel & reuse this slice of functionality?